### PR TITLE
8283694: Improve bit manipulation and boolean to integer conversion operations on x86_64

### PR DESCRIPTION
**Pull request by merykitty**
Hi, this patch improves some operations on x86_64:

- Base variable scalar shifts have bad performance implications and should be replaced by their bmi2 counterparts if possible:
  + Bounded operands
  + Multiple uops both in fused and unfused domains
  + May result in flag stall since the operations have unpredictable flag output

- Flag to general-purpose registers operation currently uses `cmovcc`, which requires set up and 1 more spare register for constant, this could be replaced by set, which transforms the sequence:

        xorl dst, dst
        sometest
        movl tmp, 0x01
        cmovlcc dst, tmp

        into:

        xorl dst, dst
        sometest
        setbcc dst

This sequence does not need a spare register and without any drawbacks.
(Note: `movzx` does not work since move elision only occurs with different registers for input and output)

- Some small improvements:
  + Add memory variances to `tzcnt` and `lzcnt`
  + Add memory variances to `rolx` and `rorx`
  + Add missing `rolx` rules (note that `rolx dst, imm` is actually `rorx dst, size - imm`)

The speedup can be observed for variable shift instructions

        Before:
        Benchmark               (size)  Mode  Cnt   Score   Error  Units
        Integers.shiftLeft         500  avgt    5   0.836 ± 0.030  us/op
        Integers.shiftRight        500  avgt    5   0.843 ± 0.056  us/op
        Integers.shiftURight       500  avgt    5   0.830 ± 0.057  us/op
        Longs.shiftLeft            500  avgt    5   0.827 ± 0.026  us/op
        Longs.shiftRight           500  avgt    5   0.828 ± 0.018  us/op
        Longs.shiftURight          500  avgt    5   0.829 ± 0.038  us/op

        After:
        Benchmark               (size)  Mode  Cnt   Score   Error  Units
        Integers.shiftLeft         500  avgt    5   0.761 ± 0.016  us/op
        Integers.shiftRight        500  avgt    5   0.762 ± 0.071  us/op
        Integers.shiftURight       500  avgt    5   0.765 ± 0.056  us/op
        Longs.shiftLeft            500  avgt    5   0.755 ± 0.026  us/op
        Longs.shiftRight           500  avgt    5   0.753 ± 0.017  us/op
        Longs.shiftURight          500  avgt    5   0.759 ± 0.031  us/op

For `cmovcc 1, 0`, I have not been able to create a reliable microbenchmark since the benefits are mostly regarding register allocation.

Thank you very much.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8283694](https://bugs.openjdk.java.net/browse/JDK-8283694): Improve bit manipulation and boolean to integer conversion operations on x86_64

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [228427b8](https://git.openjdk.java.net/jdk/pull/7968/files/228427b8d1477292e5e37b6a5c179c9111112256)
 * [Dean Long](https://openjdk.java.net/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7968/head:pull/7968` \
`$ git checkout pull/7968`

Update a local copy of the PR: \
`$ git checkout pull/7968` \
`$ git pull https://git.openjdk.java.net/jdk pull/7968/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7968`

View PR using the GUI difftool: \
`$ git pr show -t 7968`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7968.diff">https://git.openjdk.java.net/jdk/pull/7968.diff</a>

</details>

### CODE REVIEW

**Comment by iklam** [comment]
Is it possible to change the RFE title to something more specific? The current title "Improve several operations on x86" makes it difficult to look up similar issues in the future.

Also, does any of the operations apply on 32-bit x86? If not, the RFE title should say x86_64.

**Comment by merykitty** [comment]
I have adjusted the title to be more specific and accurate. Thank you very much.

**Comment by iklam** [comment]
> I have adjusted the title to be more specific and accurate. Thank you very much.

Thanks for fixing the title. I am not familiar with the code so I'll let others review it.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86_64.ad:9438]: Why this change to type of constant?

**Review by vnkozlov — COMMENTED** [review]
Few comments.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (merykitty) [reply src/hotspot/cpu/x86/x86_64.ad:9061]: Hi, the check is to match the predicates of `rReg_rReg` and `mem_rReg` versions so that the ADLC can correctly mark the latter to be the cisc-spill of the former. And the predicate of `salI_mem_rReg` is to prevent `LShiftI (LoadI mem) imm` from being matched by the variable shift bmi2 rule.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (merykitty) [reply src/hotspot/cpu/x86/x86_64.ad:9438]: For other shift nodes, the `Ideal` method clips the constant shift count to be in the correct range so we can match `immI8` with it. The `RShiftLNode` does not have an `Ideal` method so we have to do that in the backend. Previously, constant shifts that out-of-bounds for 8-bit immediate still match with the variable shift rule, but as I exclude constant shifts from `sarL_rReg_rReg`, this reveals.
Thank you very much.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (merykitty) [reply src/hotspot/cpu/x86/x86_64.ad]: I have changed the rule to use `ins_cost` instead for the prevention. Thanks.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (vnkozlov) [reply src/hotspot/cpu/x86/x86_64.ad:9434]: okay.

**Review by vnkozlov — COMMENTED** [review]
Thank you for answering my questions. Let me test it before approval.

**Review by vnkozlov — APPROVED** [review]
Testing passed.

**Comment by vnkozlov** [comment]
You need second review.

**Comment by merykitty** [comment]
Hi, may I have a second review for this patch, please? Thank you very much.

File: src/hotspot/cpu/x86/x86_64.ad
Comment (merykitty) [reply src/hotspot/cpu/x86/x86_64.ad]: Yes you are right, the change reduce 1 cycle of latency but may require another register so I reverted it. Thanks a lot.

**Review by dean-long — APPROVED** [review]

**Comment by merykitty** [comment]
Thank you very much for the reviews.
/integrate

File: src/hotspot/cpu/x86/x86_64.ad
Comment (dean-long) [reply src/hotspot/cpu/x86/x86_64.ad]: Does this pattern give any improvement?
```
testl $src, $src
movl $dst, 0
setnz $dst
```

Comment (merykitty) [reply src/hotspot/cpu/x86/x86_64.ad]: `movl r, 0` is not a zero idiom, so a partial register write later of `setcc` would lead to register stall when the value is read.

**Comment by vnkozlov** [comment]
I will run testing of latest versiuon before sponsoring.

**Comment by vnkozlov** [comment]
Testing results are good.

**Pull request closed without merging**
